### PR TITLE
Adding GPT-2 model

### PR DESCRIPTION
**Pull request by neginraoof**
# GPT-2

## Use-cases
Transformer-based language model for text generation.

## Description
[GPT-2](https://openai.com/blog/better-language-models/) is a large transformer-based language model with a simple objective: predict the next word, given all of the previous words within some text.

## Model

 |Model        |Download  |Checksum| Download (with sample test data)|ONNX version|Opset version|Accuracy |
|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
|GPT-2       |[522.81 MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2/model/model.onnx) | [MD5](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2/model/gpt2-md5.txt)| [438.3 MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2/model/GPT-2.tar.gz)| 1.6 | 10 |mAP of [0.024](https://docs.google.com/spreadsheets/d/1sryqufw2D0XlUH4sq3e9Wnxu5EAQkaohzrJbd5HdQ_w/edit#gid=0)|

## Inference
The script for ONNX model conversion and ONNXRuntime inference is [here](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2/GPT2-export.py)

### Input to model
Sequence of words as a string. Example: "Here is some text to encode : Hello World", tokenized by Byte-Pair-Encoding.
**input_ids**: Indices of input tokens in the vocabulary. It's a long tensor of dynamic shape (batch_size, sequence_length).

### Preprocessing steps
Use ```tokenizer.encode()``` to encode the input text:
```python
text = "Here is some text to encode : Hello World"
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
tokens_tensor = torch.tensor([torch.tensor(tokenizer.encode(text))])
```

### Output of model
**last_hidden_state**: Sequence of hidden-states at the last layer of the model. It's a float tensor of size (batch_size, sequence_length, hidden_size).
**past**: pre-computed hidden-states. It's a list of tensors (key and values in the attention blocks) of size (batch_size, num_heads, sequence_length, sequence_length), one per each layer.

Output of this model is the tuple (last_hidden_state, past)

Note that output_hidden_states=False and output_attentions=False in the PretrainedConfig configs.

### Postprocessing steps
```python
outputs = model(input_ids)
last_hidden_states = outputs[0]
```
<hr>

## Dataset (Train and validation)
The original model from OpenAI is pretrained on a dataset of [8 million web pages](https://openai.com/blog/better-language-models).
The pretrained model is referenced in  [huggingface/transformers](https://github.com/huggingface/transformers/blob/master/transformers/modeling_gpt2.py) repository as a causal (unidirectional) transformer pre-trained using language modeling on a very large corpus of ~40 GB of text data.
https://s3.amazonaws.com/models.huggingface.co/bert/gpt2-pytorch_model.bin

<hr>

## Validation accuracy
Metric and benchmarking details are provided by HuggingFace in this [post](https://medium.com/huggingface/benchmarking-transformers-pytorch-and-tensorflow-e2917fb891c2).
<hr>

## Publication/Attribution
Alec Radford, Jeffrey Wu, Rewon Child, David Luan, Dario Amodei, andIlya Sutskever. Language Models are Unsupervised Multitask Learners. 2019.

## References
This model is converted directly from [huggingface/transformers](https://github.com/huggingface/transformers/blob/master/transformers/modeling_gpt2.py)
<hr>

## Contributors
Negin Raoof
<hr>

## License
Apache 2.0 License
<hr>

### CODE REVIEW

**Comment by prasanthpul** [comment]
Please put the model files in Git LFS

File: text/machine_comprehension/gpt-2/README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ |Model        |Download  |Checksum| Download (with sample test data)|ONNX version|Opset version|Accuracy |
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
+|GPT-2       |[522.81 MB](https://onnxzoo.blob.core.windows.net/models/opset_10/GPT2/model.onnx) | [MD5](https://onnxzoo.blob.core.windows.net/models/opset_10/GPT2/gpt2-md5.txt)| [438.3 MB](https://onnxzoo.blob.core.windows.net/models/opset_10/GPT2/GPT2.tar.xz)| 1.6 | 10 |mAP of [0.024](https://docs.google.com/spreadsheets/d/1sryqufw2D0XlUH4sq3e9Wnxu5EAQkaohzrJbd5HdQ_w/edit#gid=0)|
+
+
+## Inference
+
+### Input to model
+Sequence of words as a string. Example: "Here is some text to encode : Hello World", tokenized by Byte-Pair-Encoding.
+**input_ids**: Indices of input tokens in the vocabulary.
```
Comment (BowenBao) [inline text/machine_comprehension/gpt-2/README.md]: Please also provide the expected input/output shape and dtype as well.

File: text/machine_comprehension/gpt-2/README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+**past**: precomputed hidden-states
+Other optional inputs to the model are:
```
Comment (BowenBao) [inline text/machine_comprehension/gpt-2/README.md]: Based on configurations the original model could have optional inputs and outputs.
On the other hand, in this page the uploaded ONNX model should have already chosen a specific configuration, and has fixed number of inputs/outputs. Let's note them down here.

**Comment by EmmaNingMS** [comment]
Could you also upload the whole script including conversion and inference?

File: text/machine_comprehension/gpt-2/README.md (hunk 2 above)
Comment (neginraoof) [reply text/machine_comprehension/gpt-2/README.md]: Sure. Makes sense.

**Comment by neginraoof** [comment]
> Could you also upload the whole script including conversion and inference?

@EmmaNingMS I added that to the repo. Do you want me to reference it in the README file?

**Comment by EmmaNingMS** [comment]
> Could you also upload the whole script including conversion and inference?
> 
> @EmmaNingMS I added that to the repo. Do you want me to reference it in the README file?

Yes, pls.

**Comment by neginraoof** [comment]
@prasanthpul I think this is ready. Please review.
cc @EmmaNingMS

File: .gitattributes [hunk 3]
```diff
@@ -6,3 +6,4 @@
 *.pkl filter=lfs diff=lfs merge=lfs -text
 *.tar.gz filter=lfs diff=lfs merge=lfs -text
 *.tgz filter=lfs diff=lfs merge=lfs -text
+*.tar.xz filter=lfs diff=lfs merge=lfs -text
```
Comment (prasanthpul) [inline .gitattributes]: so far, all the other models are tar.gz. for consistency, would be better to keep the same. any particular reason to use tar.xz?

Comment (neginraoof) [reply .gitattributes]: No. I just uploaded tar.gz and updated the PR

File: text/machine_comprehension/gpt-2/README.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+## References
+This model is converted directly from [huggingface/transformers](https://github.com/huggingface/transformers)
+<hr>
+
+## Contributors
+Negin Raoof
+<hr>
+
+## License
+Apache License
```
Comment (prasanthpul) [inline text/machine_comprehension/gpt-2/README.md]: More specifically it's the Apache 2.0 license, right? (since that's what hugging face uses)

File: text/machine_comprehension/gpt-2/README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+## Dataset (Train and validation)
+The original model from OpenAI is pretrained on a dataset of [8 million web pages](https://openai.com/blog/better-language-models).
+The pretrained model is referenced in  [huggingface/transformers](https://github.com/huggingface/transformers/blob/master/transformers/modeling_gpt2.py) repository as a causal (unidirectional) transformer pre-trained using language modeling on a very large corpus of ~40 GB of text data.
+https://s3.amazonaws.com/models.huggingface.co/bert/gpt2-pytorch_model.bin
+
+<hr>
+
+## Validation accuracy
+Metric and benchmarking details are provided by Huggigngface in this [post](https://medium.com/huggingface/benchmarking-transformers-pytorch-and-tensorflow-e2917fb891c2).
```
Comment (prasanthpul) [inline text/machine_comprehension/gpt-2/README.md]: typo: huggigngface -> HuggingFace

File: text/machine_comprehension/gpt-2/README.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+Transformer-based language model for text generation.
+
+## Description
+[GPT-2](https://openai.com/blog/better-language-models/) is a large transformer-based language model with a simple objective: predict the next word, given all of the previous words within some text.
+
+## Model
+
+ |Model        |Download  |Checksum| Download (with sample test data)|ONNX version|Opset version|Accuracy |
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
+|GPT-2       |[522.81 MB](https://onnxzoo.blob.core.windows.net/models/opset_10/GPT2/model.onnx) | [MD5](https://onnxzoo.blob.core.windows.net/models/opset_10/GPT2/gpt2-md5.txt)| [438.3 MB](https://onnxzoo.blob.core.windows.net/models/opset_10/GPT2/GPT-2.tar.gz)| 1.6 | 10 |mAP of [0.024](https://docs.google.com/spreadsheets/d/1sryqufw2D0XlUH4sq3e9Wnxu5EAQkaohzrJbd5HdQ_w/edit#gid=0)|
```
Comment (prasanthpul) [inline text/machine_comprehension/gpt-2/README.md]: links should be to the github file

Comment (neginraoof) [reply text/machine_comprehension/gpt-2/README.md]: I was following the same pattern we had for previous models like Faster RCNN https://github.com/onnx/models/blob/8d50e3f598e6d5c67c7c7253e5a203a26e731a1b/vision/object_detection_segmentation/faster-rcnn/README.md
But I can change that. Do you mean both model and md5 files?

Comment (prasanthpul) [reply text/machine_comprehension/gpt-2/README.md]: for fasterrcnn, the model was not put in Git LFS because it wasnt enabled back then. take a look at BERT for a more recent example after LFS was enabled,

**Comment by neginraoof** [comment]
@prasanthpul Can we merge this please?
